src/libsyntax/ext/expand.rs

```diff
@@ -332,13 +357,66 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
         if self.monotonic {
             let err_count = self.cx.parse_sess.span_diagnostic.err_count();
             let mark = self.cx.current_expansion.mark;
-            self.cx.resolver.visit_expansion(mark, &result.0);
+            self.cx.resolver.visit_expansion(mark, &result.0, derives);
             self.cx.resolve_err_count += self.cx.parse_sess.span_diagnostic.err_count() - err_count;
         }
 
         result
     }
 
+    fn resolve_invoc(&mut self, invoc: &mut Invocation, scope: Mark, force: bool)
+                     -> Result<Option<Rc<SyntaxExtension>>, Determinacy> {
+        let (attr, traits, item) = match invoc.kind {
+            InvocationKind::Bang { ref mac, .. } => {
+                return self.cx.resolver.resolve_macro(scope, &mac.node.path, force).map(Some);
+            }
+            InvocationKind::Attr { attr: None, .. } => return Ok(None),
+            InvocationKind::Derive { name, span, .. } => {
+                let path = ast::Path::from_ident(span, Ident::with_empty_ctxt(name));
+                return self.cx.resolver.resolve_derive_macro(scope, &path, force).map(Some);
+            }
+            InvocationKind::Attr { ref mut attr, ref traits, ref mut item } => (attr, traits, item),
+        };
+
+        let (attr_name, path) = {
+            let attr = attr.as_ref().unwrap();
+            (attr.name(), ast::Path::from_ident(attr.span, Ident::with_empty_ctxt(attr.name())))
+        };
+
+        let mut determined = true;
+        match self.cx.resolver.resolve_macro(scope, &path, force) {
+            Ok(ext) => return Ok(Some(ext)),
+            Err(Determinacy::Undetermined) => determined = false,
+            Err(Determinacy::Determined) if force => return Err(Determinacy::Determined),
+            _ => {}
+        }
+
+        for &(name, span) in traits {
```

Comment:
I don't quite understand what's happening here. For each InvocKind::Attr, we have a possible instance of an insert attribute, then we loop through the traits seeing if it belongs to any invocation? And if it does, we simply add it to a dummy item, and mark it as used?